net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+ */
+static void libfab_ep_release(struct m0_ref *ref)
+{
+	struct m0_net_end_point *nep;
+	struct m0_fab__ep       *ep;
+	struct m0_fab__tm       *tm;
+
+	nep = container_of(ref, struct m0_net_end_point, nep_ref);
+	ep = libfab_ep_net(nep);
+	tm = nep->nep_tm->ntm_xprt_private;
+
+	m0_nep_tlist_del(nep);
+	libfab_ep_param_free(ep, tm);
+	
+}
+
+/**
+ * Generate unique key for memory registration.
+ */
+static uint64_t libfab_mr_keygen(void)
+{
+	uint64_t key = FAB_MR_KEY + mr_key_idx;
+	mr_key_idx++;
+	return key;
+}
+
+/**
+ * Read single event from event queue.
+ */
+static int libfab_check_for_event(struct fid_eq *eq)
+{
+	struct fi_eq_cm_entry entry;
+	uint32_t              event = 0;
+	ssize_t               rd;
+
+	rd = fi_eq_read(eq, &event, &entry, sizeof(entry), 0);
+	if (rd == -FI_EAVAIL) {
+		struct fi_eq_err_entry err_entry;
+		fi_eq_readerr(eq, &err_entry, 0);
+		M0_LOG(M0_ERROR, "%s %s\n", fi_strerror(err_entry.err),
+		       fi_eq_strerror(eq, err_entry.prov_errno,
+				      err_entry.err_data,NULL, 0));
+		return rd;
+	}
+
+	return event;
+}
+
+/**
+ * A helper function to read the entries from a completion queue
+ * If success, returns the number of entries read
+ * else returns the negative error code
+ */
+static int libfab_check_for_comp(struct fid_cq *cq, struct m0_fab__buf **ctx,
+				 m0_bindex_t *len, uint64_t *data)
+{
+	struct fi_cq_data_entry entry[FAB_MAX_COMP_READ];
+	struct fi_cq_err_entry  err_entry;
+	uint64_t                wr_cqdata = FI_REMOTE_WRITE | FI_REMOTE_CQ_DATA;
+	int                     i;
+	int                     ret;
+	
+	ret = fi_cq_read(cq, entry, FAB_MAX_COMP_READ);
+	if (ret > 0) {
+		for (i = 0; i < ret; i++) {
+			ctx[i] = (struct m0_fab__buf *)entry[i].op_context;
+			if (len != NULL)
+				len[i] = entry[i].len;
+			if (data != NULL)
+				data[i] = ((entry[i].flags & wr_cqdata)) ? 
+				entry[i].data : 0;
+		}
+	}
+	else if (ret != -FI_EAGAIN) {
+		fi_cq_readerr(cq, &err_entry, 0);
+		M0_LOG(M0_ERROR, "%s %s\n", fi_strerror(err_entry.err),
+		       fi_cq_strerror(cq, err_entry.prov_errno,
+				      err_entry.err_data, NULL, 0));
+	}
+
+	return ret;
+}
+
+/**
+ * Finalises the transfer machine.
+ *
+ * This is called from the normal finalisation path (ma_fini(), ma_stop()) and
+ * in error cleanup case during initialisation (tm_init()).
+ */
+static void libfab_tm_fini(struct m0_net_transfer_mc *tm)
+{
+	struct m0_fab__tm *ma = tm->ntm_xprt_private;
+	int                rc = 0;
+
+	if (!ma->ftm_shutdown) {
+		while(1) {
+			libfab_tm_lock(ma);
+			if (m0_mutex_trylock(&ma->ftm_evpost) != 0) {
+				libfab_tm_unlock(ma);
+			} else
+				break;
+		}
+		m0_mutex_unlock(&ma->ftm_evpost);
+		m0_mutex_lock(&ma->ftm_endlock);
+		ma->ftm_shutdown = true;
+		m0_mutex_unlock(&ma->ftm_endlock);
+
+		libfab_tm_buf_done(ma);
+
+		rc = libfab_tm_param_free(ma);
+		if (rc != FI_SUCCESS)
+			M0_LOG(M0_ERROR, "libfab_tm_param_free ret=%d", rc);
+
+		m0_mutex_fini(&ma->ftm_endlock);
```

Comment:
How can you be sure that, at this point, the "ma->ftm_endlock" is not used by poller thread?

Reply:
The function call to libfab_tm_param_free() ensures that the poller thread has been finalised